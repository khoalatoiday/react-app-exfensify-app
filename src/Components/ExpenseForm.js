/*
    - Tạo Form cho add Expense và edit Expense
    - Sử dụng regular expression(regux101)
    - Tạo lịch với react-dates library và format kiểu của thời gian với Object Moment
*/

import moment from "moment";
import React from "react";
import { SingleDatePicker } from "react-dates"; // sử dụng với Object moment

import "react-dates/initialize";


export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense? props.expense.description : '',
      note: props.expense? props.expense.note : "",
      amount: props.expense? (props.expense.amount/100).toString() : '',
      createAt: props.expense? moment(props.expense.createAt) : moment(),
      focused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description,
    }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note,
    }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // sử dụng rerular expression để tạo điều kiện là phải là 1 số có nhiều nhất 2 số 0 sau dấu phẩy
    if (!amount || amount.match(/^\d*(\.\d{0,2})?$/gm)) {
      this.setState(() => ({
        amount,
      }));
    }
  };

  onDateChange = (createAt) => {
    if (createAt) {
      this.setState(() => ({
        createAt,
      }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      focused,
    }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide amount and desciption",
      }));
    } else {
      this.setState(() => ({
        error: "",
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          {this.state.error && (
            <p>
              <strong>{this.state.error}</strong>
            </p>
          )}
          <input
            type="text"
            placeholder="description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createAt} // khởi tạo date
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <br />
          <textarea
            placeholder="Add a note for your expense(optional)"
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
