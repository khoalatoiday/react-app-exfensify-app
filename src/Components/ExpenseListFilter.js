// lí thuyết HOC và cơ chế connect xem lại ở ExpensesList.js, HigherOrderComponent.js
// dispatch from connected RC
/*
    - Tất cả các RC được connect với redux store đều có 1 props là dispatch giúp dispatch action
    - Connect RC mỗi khi dispatch action -> sẽ thay đổi state trong redux store -> re-render lại -> khá giống AJAX
*/
import React from "react";
import { connect } from "react-redux";
import {
  setTextForFilter,
  storeByAmount,
  storeByDate,
  setEndDate,
  setStartDate,
} from "../actions/filter";
import { DateRangePicker } from "react-dates";

// props của ExpenseListFilter được lấy từ props của HOC chứa nó, export phục vụ việc test
export class ExpenseListFilter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      focused: null,
    };
    this.onDatesChange = this.onDatesChange.bind(this);
    this.onFocusChange = this.onFocusChange.bind(this);
    this.onTextChange = this.onTextChange.bind(this)
    this.onSortChange = this.onSortChange.bind(this)
  }

  onDatesChange({ startDate, endDate }) {
    this.setState(() => {
      this.props.setStartDate(startDate);
      this.props.setEndDate(endDate);
    });
  }

  onFocusChange(focused) {
    this.setState(() => ({
      focused,
    }));
  }

  onTextChange(e){
    this.props.setTextForFilter(e.target.value);
  }

  onSortChange (e){
    const choosen = e.target.value; // lấy giá trị của option được chọn
    if (choosen === "date") {
      this.props.storeByDate();
    } else if (choosen === "amount") {
      this.props.storeByAmount();
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filter.text}
          onChange={this.onTextChange}
        />
        <select
          value={this.props.filter.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDateId="your_unique_start_date_id"
          endDateId="your_unique_end_date_id"
          startDate={this.props.filter.startDate}
          endDate={this.props.filter.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateStoreToProps = (state) => {
  return {
    filter: state.filter,
  };
};

const mapDispatchToProps = (dispatch) =>{
  return {
    setTextForFilter: (text) => dispatch(setTextForFilter(text)),
    storeByAmount: () => dispatch(storeByAmount()),
    storeByDate: () => dispatch(storeByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
  }
}

export default connect(mapStateStoreToProps, mapDispatchToProps)(ExpenseListFilter);
