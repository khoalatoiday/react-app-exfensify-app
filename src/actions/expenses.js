const uuid = require("uuid");

const ADD_EXPENSES = ({
  description = "",
  note = "",
  amount = 0,
  createAt = 0,
} = {}) => ({
  type: "ADD_EXPENSES",
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createAt,
  },
});

const REMOVE_EXPENSES = ({ id } = {}) => ({
  type: "REMOVE_EXPENSES",
  expense: {
    id,
  },
});

const EDIT_EXPENSE = (id, updates) => ({
  // updates là một Object
  type: "EDIT_EXPENSES",
  id,
  updates,
});

export {ADD_EXPENSES, REMOVE_EXPENSES, EDIT_EXPENSE}