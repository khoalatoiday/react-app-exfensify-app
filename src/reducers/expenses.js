//Expenses Reducer
const expensesDefaultValueState = [];
// tại node đầu tiên thì không có state -> set up state = DefaultState
export default (state = expensesDefaultValueState, action) => {
  switch (action.type) {
    case "ADD_EXPENSES":
      return [...state, action.expense]; // new array với toàn bộ phần tử trong state + action.expense
    case "REMOVE_EXPENSES":
      return state.filter((expense) => action.expense.id != expense.id); // trả về một dãy thỏa mãn điều kiện
    case "EDIT_EXPENSES":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        }else{
          return {
            ...expense
          }
        }
      });
    default:
      return state;
  }
};
