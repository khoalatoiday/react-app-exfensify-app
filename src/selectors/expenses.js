import moment from "moment";

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  // retrun a new array thỏa mãn các điều kiện
  return expenses
    .filter((expense) => {
      const createAtMoment = moment(expense.createAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createAtMoment, "day")
        : true; // moment query doc, startDate có <= thời gian được tạo của expense
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createAtMoment, "day")
        : true; // endDate có >= thời gian được tạo của expense
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((firstEle, secondEle) => {
      if (sortBy === "date") {
        return firstEle.createAt < secondEle.createAt ? 1 : -1; // nếu xảy ra điều kiện thì đẩy secondEle lên trước -> sắp xếp giảm dần
      }
      if (sortBy === "amount") {
        return firstEle.amount > secondEle.amount ? 1 : -1; // nếu xảy ra điều kiện thì đẩy secondEle lên trước -> sắp xếp tăng dần
      }
    });
};

export { getVisibleExpenses as default };
