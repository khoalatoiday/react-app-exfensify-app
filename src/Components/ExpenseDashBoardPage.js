import React from "react";
import ExpensesList from "./ExpensesList";
import ExpenseListFilter from "./ExpenseListFilter";
import SummaryExpenses from "./SummaryExpenses";
const ExpenseDashBoardPage = () => (
  <div>
    <ExpenseListFilter />
    <ExpensesList />
    <SummaryExpenses />
    <div>Hello World</div>
  </div>
);

export { ExpenseDashBoardPage as default };
