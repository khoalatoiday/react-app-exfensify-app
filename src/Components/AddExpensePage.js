import React from "react";
import ExpenseForm from "./ExpenseForm";
import { ADD_EXPENSES } from "../actions/expenses";
import { connect } from "react-redux";

// ta muốn RC này có thể dispatch action thay đổi data trong redux store => phải đưa về 1 connected RC
// export phục vụ việc test
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.addExpense(expense)
    this.props.history.push("/") // sd history API để redirect về page mong muốn 
  };
  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

// dispatch action và đưa vào props -> gọi props là dispatch action
const mapDispatchToProps = (dispatch)=>({
  addExpense: (expense) => dispatch(ADD_EXPENSES(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
