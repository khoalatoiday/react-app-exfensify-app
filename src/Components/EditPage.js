import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { EDIT_EXPENSE, REMOVE_EXPENSES } from "../actions/expenses";

// export phục vụ việc test
export class EditPage extends React.Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this)
    this.removeExpense = this.removeExpense.bind(this)
  }

  onSubmit(expense) {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  }

  removeExpense() {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  }

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    );
  }
}

// props là props hiện tại của EditPage
const mapPropsStateToRC = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    editExpense: (id, expense) => dispatch(EDIT_EXPENSE(id, expense)),
    removeExpense: (data) => dispatch(REMOVE_EXPENSES(data)),
  };
};

export default connect(mapPropsStateToRC, mapDispatchToProps)(EditPage);
