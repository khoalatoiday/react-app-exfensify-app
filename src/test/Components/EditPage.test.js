import React from "react";
import { shallow } from "enzyme";
import { EditPage } from "../../Components/EditPage";
import expenses from "../fixtures/expense";

let wrapper, history, editExpense, removeExpense

beforeEach(() => {
  editExpense = jest.fn(); // spy function
  removeExpense = jest.fn();
  history = { push: jest.fn() }; // spy function
  wrapper = shallow(
    <EditPage
      history={history}
      editExpense={editExpense}
      removeExpense={removeExpense}
      expense={expenses[0]}
    />
  );
});

test("Should render EditPage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should handler Edit Expense", ()=>{
    wrapper.find("ExpenseForm").prop("onSubmit")({description: "demo"})
    expect(history.push).toHaveBeenLastCalledWith("/")
    expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, {description: "demo"})
})

test("Should handler Remove Expense", ()=>{
    wrapper.find("button").simulate("click")
    expect(history.push).toHaveBeenLastCalledWith("/")
    expect(removeExpense).toHaveBeenLastCalledWith({
        id: expenses[0].id
    })
})