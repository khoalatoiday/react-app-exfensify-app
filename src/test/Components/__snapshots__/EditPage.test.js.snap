// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditPage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1000,
        "createAt": -345600000,
        "description": "first",
        "id": "1",
        "note": "note",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
